File: py/y2024/day3.py
import sys
import re

def parse(input: str):
    return input

def part1(input: str):
    sum = 0
    for match in re.findall(r"mul\((\d+),(\d+)\)", input):
        a, b = map(int, match)
        sum += a * b
    return sum

def part2(input: str):
    # print("before: ", len(input))
    input = re.sub(r"don't\(\)(.*?)($|do\(\))", "", input, flags=re.DOTALL)
    # print("after:  ", len(input))
    print(input)
    sum = 0
    for match in re.findall(r"mul\((\d+),(\d+)\)", input):
        a, b = map(int, match)
        sum += a * b
    return sum

if __name__ == "__main__":
    input = sys.stdin.read()
    result = part2(input)
    print(result)

File: vite/src/y2024/day3/day3.ts
export function part1(input: string) {
    const matches = input.matchAll(/mul\((\d+),(\d+)\)/g)
    let sum = 0;
    for (const match of matches ?? []) {
        const a = Number(match[1])
        const b = Number(match[2])
        sum += a * b
    }
    return sum
}

export function part2(input: string) {
    const replaced = input.replaceAll(/don't\(\)(.*?)do\(\)/g, "don't()do()")
    return part1(replaced)
}